### PR TITLE
Fix a false positive for `simplify-boolean-expression` when multiple values inferred

### LINKED ISSUE #7626: False positive ``simplify-boolean-expression`` when boolean variables are modified after their intialization in a for loop
### Bug description

For the following code:

```python
# pylint: disable=missing-module-docstring

def some_function() -> bool:
    '''a function

    Returns:
        True
    '''
    flag_a = False
    flag_b = False
    for word in ['alfa', 'bravo']:
        if 'b' not in word:
            flag_a = True
        else:
            flag_b = True
    multiple = (flag_a and flag_b) or returns_false()  # <- error here
    return multiple

def returns_false() -> bool:
    '''a different function

    Returns:
        False, but pylint probably doesn't know that
    '''
    return False


def main() -> None:
    '''entry point
    '''
    print(repr(some_function()))  # True

if __name__ == '__main__':
    main()
```

...pylint emits:
```
R1709: Boolean expression may be simplified to returns_false() (simplify-boolean-expression)
```

This statement is incorrect: `some_function()` always returns `True`.

Note: removing ...`or returns_false()` causes the error to disappear.

### Configuration

_No response_

### Command used

```shell
pylint --rcfile=/dev/null thecodeabove.py
```

### Pylint output

```shell
************* Module thecodeabove
thecodeabove.py:17:4: R1709: Boolean expression may be simplified to returns_false() (simplify-boolean-expression)

------------------------------------------------------------------
Your code has been rated at 9.38/10 (previous run: 9.38/10, +0.00)
```

### Expected behavior

Pylint should not emit R1709/simplify-boolean-expression, since it is mistaken.

### Pylint version

```shell
pylint 2.15.4
astroid 2.12.11
Python 3.8.12 (default, Jan 17 2022, 13:29:00)
[Clang 10.0.1 (clang-1001.0.46.4)]
```

### OS / Environment

macOS 10.14

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description
If multiple values are possible for a constant, we should silence the refactoring messages that claim to offer simplifications, since the inferred value is not reliable.

Adds a keyword-only `compare_constants` argument to `safe_infer` to accomplish this in a reusable way.

Closes #7626

I'm wondering if @areveny you might wish to have a look, as I saw #5227? No pressure.

### CODE REVIEW

File: pylint/checkers/utils.py
Comment (areveny) [inline pylint/checkers/utils.py:1385]:
```suggestion
            if (
                compare_constants
                and isinstance(inferred, nodes.Const)
                and isinstance(value, nodes.Const)
                and inferred.value != value.value
            ):
                return None
```

I think you could write it like this and remove the logic above since `value` isn't mutated.

**Review by areveny — APPROVED** [review]
Looks good to me! Left one suggestion. Requesting another approver.

I wonder if there's anywhere else `compare_constants=True` could be used but after perusing usages of `safe_infer` no other applications are immediately clear.

File: doc/whatsnew/fragments/7626.false_positive
Comment (Pierre-Sassoulas) [inline doc/whatsnew/fragments/7626.false_positive:4]: This one is an internal change, but we might want to add two fragments as it's indeed valuable to know for plugin developers.
```suggestion
```

File: tests/functional/t/ternary.py
Comment (jacobtylerwalls) [reply tests/functional/t/ternary.py:2]: The previous tests were wrong IMO, they caused the same false positive as reported in the issue. If something is uninferable, we should not assume it's truthy!

I agree I can make this more explicit with a test.

Comment (Pierre-Sassoulas) [reply tests/functional/t/ternary.py:2]: > If something is uninferable, we should not assume it's truthy!

Agree. Maybe we can keep only one test for this scenario instead of what we had before ?

**Review by Pierre-Sassoulas — DISMISSED** [review]
We could re-add the old test for uninferable value and fix it, but I'll let you decide :)

**Comment by jacobtylerwalls** [comment]
Ah, I think I follow you now. I've got versions of the original cases, plus two new cases.

**Review by Pierre-Sassoulas — APPROVED** [review]
👌

**Comment by jacobtylerwalls** [comment]
Thanks for the reviews!

**Comment by Pierre-Sassoulas** [comment]
This is going to be released in pylint 2.15.5 :)

File: pylint/checkers/utils.py
Comment (DanielNoord) [inline pylint/checkers/utils.py:1376]: I'm not sure how the compiler handle this but since this is called so often we might want to "optimize" the if's here. If we use `if isinstance(inferred, ...)` and then go down a level we can use `elifs` to make sure we aren't doing `if` statement that we know will never be true.

Not big enough to create a follow-up but I guess something to consider for future PRs that touch this code.

Comment (Pierre-Sassoulas) [reply pylint/checkers/utils.py:1376]: Good point. I'm not sure we'll be able to keep that in mind if we don't do the follow-up immediately though.

Comment (jacobtylerwalls) [reply pylint/checkers/utils.py:1376]: > I'm not sure how the compiler handle this

`and` statements don't evaluate the right hand side if the left side tests false:
```python
>>> not 1 and (1 / 0)
False
```

Do you have a suggested optimization that would short-circuit more?

Comment (DanielNoord) [reply pylint/checkers/utils.py:1376]: Since it's merged I can't create an actual suggestion, but:
```python
isinstance(inferred, nodes.Const):
    if compare_constants ...
        ...
elif isinstance(inferred, nodes.FunctionDef):
    if inferred.args.args is not None:
      ...
```

Comment (jacobtylerwalls) [reply pylint/checkers/utils.py:1376]: Ah. Perhaps. But we would need a microbenchmark for how many times we call `compare_constants`, as it's probably way faster to check the boolean, right?

Comment (DanielNoord) [reply pylint/checkers/utils.py:1376]: Yeah, you're right!

**Pull request merged**
